### PR TITLE
 login: XGI Z7/Z9 (XG20 core) graphic chip requires master-of-seat to…

### PR DESCRIPTION
**Pull request by fbuihuu**
… be set

This graphic chip doesn't have a DRM driver and fall back to vesa-framebuffer
driver.

Without this patch, users of such chip suddenly see their GUI broken without
any indication or reason of what happened (no error message). Hence this
regression is near to impossible to troubleshoot for end users. Such case was
reported https://bugzilla.opensuse.org/show_bug.cgi?id=1187154.

### CODE REVIEW

**Comment by fbuihuu** [comment]
Note: I can't test this myself because I don't have the HW.

**Comment by keszybz** [comment]
LGTM.

Would it be possible to get the original reports to test this patch?

**Comment by hramrach** [comment]
/tmp/autopkgtest-lxc.ml0tfnjy/downtmp/build.GaY/src/build-deb/../hwdb.d/60-master-of-seat.hwdb: 1 match groups, 1 matches, 1 properties12:57
Failed to parse: 'ID_MASTER_OF_SEAT_FROM_DATABASE=1'

**Comment by keszybz** [comment]
This needs ` ('ID_MASTER_OF_SEAT_FROM_DATABASE', Literal('1'))` in `hwdb/parse_hwdb.py`.

File: hwdb.d/60-master-of-seat.hwdb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+#    ID_MASTER_OF_SEAT_FROM_DATABASE=1
+
+#########################################
+# eXtreme Graphic Innovation (XGI)
+#########################################
+
+# Z7/Z9 (XG20 core), a rather ancient graphic chip, doesn't have a DRM driver
+# and falls back to vesafb.
+pci:v000018CAd00000020*
+ ID_MASTER_OF_SEAT_FROM_DATABASE=1
```
Comment (poettering) [inline hwdb.d/60-master-of-seat.hwdb]: Could you move this one here too:

```
# Video adapter of Parallels virtualization platform
# Seat should be synthesized for it. But there's no in-kernel driver for this
# device so matching by vid/pid.
SUBSYSTEM=="pci", ATTRS{vendor}=="0x1ab8", ATTRS{device}=="0x4005", TAG+="seat", TAG+="master-of-seat"
```

File: src/login/71-seat.rules.in [hunk 2]
```diff
@@ -29,6 +29,10 @@ SUBSYSTEM=="graphics", KERNEL=="fb[0-9]", IMPORT{cmdline}="nomodeset", TAG+="mas
 SUBSYSTEM=="pci", ENV{ID_PCI_CLASS_FROM_DATABASE}=="Display controller", \
                   ENV{DRIVER}=="", IMPORT{cmdline}="nomodeset", TAG+="seat", TAG+="master-of-seat"
 
+# Synthesize a seat for graphic devices without DRM and that fall back to fb
+# device instead.
+SUBSYSTEM=="pci", ENV{ID_MASTER_OF_SEAT_FROM_DATABASE}=="1", TAG+="seat", TAG+="master-of-seat"
```
Comment (poettering) [inline src/login/71-seat.rules.in]: the first check should be unnecessary. if the prop is set it should count.

I'd drop the "_FROM_DATABASE" suffix. We use that nly in case we have conceptually the same field from two sources. but here we don't hence no suffix please.

**Comment by poettering** [comment]
looking at the udev rules files, we actually need to set the `seat` tag on some devices, the `master-of-seat` tag on a superset of that, and the `ID_AUTOSEAT` property on a different (possibly overlapping super

This makes me thing we should probably have three props we can set from the hwdb:

1. `ID_TAG_SEAT=1`. This just sets the `seat` tag. 
2. `ID_TAG_MASTER_OF_SEAT=1`. This sets the `master-of-seat` tag.
3. `ID_AUTOSEAT=1`. This would just be import 1:1 into the device.

Now, the first item isn't too interesting right now as we have no vid/pid matching devices that actually need this right now.

The second item would be the XGI one plus the parallels one.

The third item would cover that qemu device and the plugable hub.

TLDR. please rename the property you are adding to `ID_TAG_MASTER_OF_SEAT=` in line with this. And it would be fantastic if the PR could cover the third item, but it's ok if not.

**Comment by poettering** [comment]
Hmm, and I think we should replace this rule:

```
 # HyperV currently doesn't do DRM, hence we need to synthesize for HyperV's fb device instead
SUBSYSTEM=="graphics", KERNEL=="fb[0-9]", DRIVERS=="hyperv_fb", TAG+="master-of-seat"
```

with a hwdb entry matching the pci vid/pid too for setting ID_TAG_MASTER_OF_SEAT=1. It's really the same thing. According to "modinfo hyperv_fb" the modalias string to match against is `v00001414d00005353sv00000000sd00000000bc`...

**Comment by poettering** [comment]
hmm, thinking about this more. I think tagging the PCI device with "master-of-tag" is kinda sloppy. It should always be the fb or drm device. After all, if the there are no fb nor drm drivers installed, then we should not claim the seat was graphics capable, because it isn't.

hence udev rules change should:

1. have one rule that starts with something like `SUBSYSTEM=="graphics", KERNEL=="fb[0-9]", …` and then explicitly calls the hwdb builtin, so that the modalias-based props are pulled onto the fb device, not just the PCI device.
2. a rule that then synthesizes the tags based on the results of that, but only on the fb device.

**Comment by fbuihuu** [comment]
@poettering agreed with all of your suggestions. I'm actually now wondering how tagging the PCI device with 'master-of-seat' instead of the frame-buffer device could have worked... maybe X server is smart enough to fix that up.

**Comment by fbuihuu** [comment]
@poettering I force pushed a new version. Does it look right now ?

If so I'll address the rest of your feedback.

File: src/login/71-seat.rules.in [hunk 3]
```diff
@@ -29,6 +29,11 @@ SUBSYSTEM=="graphics", KERNEL=="fb[0-9]", IMPORT{cmdline}="nomodeset", TAG+="mas
 SUBSYSTEM=="pci", ENV{ID_PCI_CLASS_FROM_DATABASE}=="Display controller", \
                   ENV{DRIVER}=="", IMPORT{cmdline}="nomodeset", TAG+="seat", TAG+="master-of-seat"
 
+# Synthesize a seat for graphic devices without DRM and that fall back to fb
+# device instead. Such HWs are listed in hwdb.
+SUBSYSTEM=="graphics", KERNEL=="fb[0-9]*", ATTRS{modalias}=="?*", IMPORT{builtin}="hwdb fb:$attr{modalias}"
+ENV{ID_TAG_MASTER_OF_SEAT}=="1", TAG+="master-of-seat"
```
Comment (poettering) [inline src/login/71-seat.rules.in:32]: please add `TAG+="seat"` here too, so that it is implied.

File: hwdb.d/60-master-of-seat.hwdb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+#    ID_TAG_MASTER_OF_SEAT=1
+
+#########################################
+# eXtreme Graphic Innovation (XGI)
+#########################################
+
+# Z7/Z9 (XG20 core), a rather ancient graphic chip, doesn't have a DRM driver
+# and falls back to vesafb.
+fb:pci:v000018CAd00000020*
+ ID_TAG_MASTER_OF_SEAT=1
```
Comment (poettering) [inline hwdb.d/60-master-of-seat.hwdb]: as suggested please move the parallels and hyperv entry here too, and do them the same way

**Comment by poettering** [comment]
looks pretty good, see comments above

File: src/login/71-seat.rules.in (hunk 3 above)
Comment (fbuihuu) [reply src/login/71-seat.rules.in:32]: hmm but this time the "seat" tag is already set since the whole thing applies to frame buffer devices (ie not the PCI ones), no ?

File: hwdb.d/60-master-of-seat.hwdb (hunk 4 above)
Comment (fbuihuu) [reply hwdb.d/60-master-of-seat.hwdb]: According to commit  8e3480ebc097a8336633436f412a9c08902962ad, there's no kernel driver for Parallels video adapter hence no framebuffer driver, which is a prerequisite for using the hwdb, no ?

File: src/login/71-seat.rules.in (hunk 3 above)
Comment (poettering) [reply src/login/71-seat.rules.in:32]: good point

File: hwdb.d/60-master-of-seat.hwdb (hunk 4 above)
Comment (poettering) [reply hwdb.d/60-master-of-seat.hwdb]: i think it tries to say that because there's no explicit driver vesafb is being used, like for your case

Comment (fbuihuu) [reply hwdb.d/60-master-of-seat.hwdb]: Ok in that case it makes sense to do the same for Parallels indeed.

Comment (fbuihuu) [reply hwdb.d/60-master-of-seat.hwdb]: But note that the rule for Parallels tagged the PCI device, ie not the framebuffer.

**Comment by fbuihuu** [comment]
@poettering new version pushed, PTAL.

Please note that I don't have the relevant HWs for testing.

**Comment by bluca** [comment]
```
 844/1201 parse-hwdb                              FAIL     3.54 s (exit status 1)

--- command ---
14:01:28 /home/runner/work/systemd/systemd/hwdb.d/parse_hwdb.py /home/runner/work/systemd/systemd/build/../hwdb.d/60-autosuspend.hwdb /home/runner/work/systemd/systemd/build/../hwdb.d/60-autosuspend-fingerprint-reader.hwdb /home/runner/work/systemd/systemd/build/../hwdb.d/60-evdev.hwdb /home/runner/work/systemd/systemd/build/../hwdb.d/60-input-id.hwdb /home/runner/work/systemd/systemd/build/../hwdb.d/60-keyboard.hwdb /home/runner/work/systemd/systemd/build/../hwdb.d/60-seat.hwdb /home/runner/work/systemd/systemd/build/../hwdb.d/60-sensor.hwdb /home/runner/work/systemd/systemd/build/../hwdb.d/70-joystick.hwdb /home/runner/work/systemd/systemd/build/../hwdb.d/70-mouse.hwdb /home/runner/work/systemd/systemd/build/../hwdb.d/70-pointingstick.hwdb /home/runner/work/systemd/systemd/build/../hwdb.d/70-touchpad.hwdb /home/runner/work/systemd/systemd/build/../hwdb.d/80-ieee1394-unit-function.hwdb hwdb.d/60-autosuspend-chromiumos.hwdb
--- stdout ---
/home/runner/work/systemd/systemd/build/../hwdb.d/60-autosuspend.hwdb: 5 match groups, 8 matches, 5 properties
/home/runner/work/systemd/systemd/build/../hwdb.d/60-autosuspend-fingerprint-reader.hwdb: 23 match groups, 216 matches, 23 properties
/home/runner/work/systemd/systemd/build/../hwdb.d/60-evdev.hwdb: 85 match groups, 110 matches, 308 properties
/home/runner/work/systemd/systemd/build/../hwdb.d/60-input-id.hwdb: 5 match groups, 5 matches, 6 properties
/home/runner/work/systemd/systemd/build/../hwdb.d/60-keyboard.hwdb: 193 match groups, 285 matches, 968 properties
Cannot parse /home/runner/work/systemd/systemd/build/../hwdb.d/60-seat.hwdb: Expected {Group:({{{{{Combine:({{{"mouse:" {"usb" ^ "bluetooth" ^ "ps2" ^ "*"} ":"} ^ {"evdev:" {"name" ^ "atkbd" ^ "input"} ":"} ^ {"id-input:" {"modalias"} ":"} ^ {"touchpad:" {"i8042" ^ "rmi" ^ "bluetooth" ^ "usb"} ":"} ^ {"joystick:" {"i8042" ^ "rmi" ^ "bluetooth" ^ "usb"} ":"} ^ {"keyboard:" {"name"} ":"} ^ {"sensor:" {"modalias"} ":"} ^ {"ieee1394-unit-function:" {"node"} ":"}} W:(0123...)}) | Combine:({{"usb" ^ "ieee1394" ^ "pci" ^ "acpi" ^ "OUI" ^ "sdio" ^ "vmbus" ^ "bluetooth"} ":" W:(0123...)})} Suppress:(LineEnd)} ^ Suppress:({Python style comment Suppress:(LineEnd)})}}... - {{{Suppress:(<SP>) Combine:({W:(ABCD..., ABCD...) - "=" - [W:(ABCD...)] - [Python style comment]}) Suppress:(LineEnd)} ^ Suppress:({<SP> Python style comment Suppress:(LineEnd)})}}... - Suppress:({LineEnd ^ stringEnd})}) ^ {Suppress:({{Python style comment Suppress:(LineEnd)}}...) - Suppress:(LineEnd)}}, found '#'  (at char 0), (line:1, col:1)
/home/runner/work/systemd/systemd/build/../hwdb.d/60-seat.hwdb: 0 match groups, 0 matches, 0 properties
/home/runner/work/systemd/systemd/build/../hwdb.d/60-seat.hwdb: no matches or props
/home/runner/work/systemd/systemd/build/../hwdb.d/60-sensor.hwdb: 152 match groups, 194 matches, 161 properties
/home/runner/work/systemd/systemd/build/../hwdb.d/70-joystick.hwdb: 2 match groups, 3 matches, 2 properties
/home/runner/work/systemd/systemd/build/../hwdb.d/70-mouse.hwdb: 113 match groups, 132 matches, 139 properties
/home/runner/work/systemd/systemd/build/../hwdb.d/70-pointingstick.hwdb: 8 match groups, 30 matches, 11 properties
/home/runner/work/systemd/systemd/build/../hwdb.d/70-touchpad.hwdb: 6 match groups, 9 matches, 6 properties
/home/runner/work/systemd/systemd/build/../hwdb.d/80-ieee1394-unit-function.hwdb: 166 match groups, 334 matches, 626 properties
hwdb.d/60-autosuspend-chromiumos.hwdb: 1 match groups, 109 matches, 1 properties
```

File: hwdb.d/60-seat.hwdb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+# device, and not the (parent) PCI device, is tagged with 'master-of-seat'.
+#
+# Allowed properties are:
+#    ID_TAG_MASTER_OF_SEAT=1
+
+#########################################
+# eXtreme Graphic Innovation (XGI)
+#########################################
+
+# Z7/Z9 (XG20 core), a rather ancient graphic chip, doesn't have a DRM driver
```
Comment (poettering) [inline hwdb.d/60-seat.hwdb:17]: maybe say that this is virtualized by some VM manager though. because as I understood this is the reason we are having the discussion

**Comment by arvidjaar** [comment]
@fbuihuu 

> I'm actually now wondering how tagging the PCI device with 'master-of-seat' instead of the frame-buffer device could have worked

Because nobody cares about actual device. All that display manager needs is `CanGraphical` property from `systemd-logind` and all that `systemd-logind` needs to set this to `true` is _some_ device with `master-of-seat` tag.

File: hwdb.d/60-seat.hwdb (hunk 5 above)
Comment (fbuihuu) [reply hwdb.d/60-seat.hwdb:17]: AFAIK XGI Z7/Z9 is a real HW unlike the 2 other cases (HyperV and Parallels).

**Comment by fbuihuu** [comment]
> ```
>  844/1201 parse-hwdb                              FAIL     3.54 s (exit status 1)
>  ```

Thanks this should be fixed now.

File: hwdb.d/60-seat.hwdb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+# and falls back to vesafb.
+fb:pci:v000018CAd00000020*
+ ID_TAG_MASTER_OF_SEAT=1
+
+#########################################
+# HyperV
+#########################################
+
+# HyperV currently doesn't do DRM, hence we need to synthesize for HyperV's fb
```
Comment (bluca) [inline hwdb.d/60-seat.hwdb:26]: There is support from v5.14: https://github.com/torvalds/linux/commit/76c56a5affeba1e163b
Does this interfere with using that? Or is it unrelated?

Comment (fbuihuu) [reply hwdb.d/60-seat.hwdb:26]: I guess this is related, however the compat entry in hwdb is still needed for a while until kernels < 5.14 are no more supported.

Comment (poettering) [reply hwdb.d/60-seat.hwdb:26]: these are the perils of encoding facts about software rather than about hardware in hwdb...

Comment (keszybz) [reply hwdb.d/60-seat.hwdb:26]: An udev rule was more flexible in this regard… But let's wait for 5.14 to be released. If it turns out that this is any problem, we can add some rule that looks at the kernel version or whatnot. We don't need to solve this now.

Comment (bluca) [reply hwdb.d/60-seat.hwdb:26]: Can we postpone this to v250 then? Autumn distros might use v5.14 and systemd v249

**Comment by keszybz** [comment]
LGTM.

Normally I'd wait until after the release, and then maybe include the patch in backports after it has had some time to bake. But this is a follow-up for #20070, which is already in v249, so let's merge this one immediately too.

File: hwdb.d/60-seat.hwdb (hunk 6 above)
Comment (fbuihuu) [reply hwdb.d/60-seat.hwdb:26]: Maybe we should pass the version of the kernel that get DRM support for the specific HW from hwdb and use that in the udev rule to decide whether the framebuffer needs `master-of-seat` tag or not.

Comment (keszybz) [reply hwdb.d/60-seat.hwdb:26]: > Can we postpone this to v250 then? Autumn distros might use v5.14 and systemd v249

I don't think so. First, at this point we don't know if 5.14 will be released with the driver, or maybe it'll get reverted or moved to staging or whatever. If if passes that first hurdle, we don't know if various distros will enable it by default. OTOH, this patch fixes real issues, so we want it sooner rather than later.

> Maybe we should pass the version of the kernel that get DRM support for the specific HW from hwdb and use that in the udev rule to decide whether the framebuffer needs master-of-seat tag or not.

Maybe. How would that look exactly?

Is it an issue if we have this rule and a working driver? Maybe it's just a bit ugly but harmless otherwise?

Comment (fbuihuu) [reply hwdb.d/60-seat.hwdb:26]: > Maybe. How would that look exactly?

Dunno maybe by introducing a new feature in udev (builtin or whatever), not sure whether it's worth the trouble though.

> Is it an issue if we have this rule and a working driver? Maybe it's just a bit ugly but harmless otherwise?

We would reintroduce issue #10435 for these devices. But given the fact that GDM got some improvements that make it less subject to this issue, having this rule and a DRM driver wouldn't be an issue I would say.

Comment (hramrach) [reply hwdb.d/60-seat.hwdb:26]: We wouldn't because incidentally in 5.14 it is planned for vesafb to also get a DRM capable driver. When this happens all of this drm/non-drm split pretty much goes away, at least for common hardware.

Comment (bluca) [reply hwdb.d/60-seat.hwdb:26]: Will we need to change this at that point? Or is it transparent?

Comment (hramrach) [reply hwdb.d/60-seat.hwdb:26]: At that point the device will get flagged twice - once as the framebuffer corresponding to a specific known device, once as DRM capable. So in theory it should just work.

Comment (bluca) [reply hwdb.d/60-seat.hwdb:26]: ah that's great to hear, thanks!

File: hwdb.d/60-seat.hwdb (hunk 5 above)
Comment (poettering) [reply hwdb.d/60-seat.hwdb:17]: @hramrach  said otherwise:

https://github.com/systemd/systemd/pull/20070#issuecomment-872055380

he said this was about virtualization. So what is it now?

File: hwdb.d/60-seat.hwdb (hunk 6 above)
Comment (poettering) [reply hwdb.d/60-seat.hwdb:26]: nah, if both drivers exist in parallel things will be racy, we need to drop the entry from the hwdb then. The problem is that it's not clear if the drm or fb device will show up first, and thus graphics apps that want to use drm preferably and fall back to fb otherwise might sometimes end up with the wrong device. If that wasn#t an issue then we could tag all fb devices out of the box, and in fact we used to do that. We dropped that only in order to fix the race. 

I mean, if there are userpace drivers for these things I guess we could just tell these packages to ship hwdb entires/udev rules for the devices they care about and then drop them from upstream. i.e. it's a property of certain software and hence should be shipped with said software. The big issue with that is just that vesafb is kinda a catchall, i.e. devices that have oo drivers of their own, hence no obvious candidate package to ship those rules.

Comment (hramrach) [reply hwdb.d/60-seat.hwdb:26]: The problem is that vesafb which is not DRM capable shows up first, and binding to it blocks the real driver from working.

With DRM support in vesafb this problem goes away when the display server uses DRM with vesafb it can switch to the real driver once available.

Recognizing that vesafb is DRM capable and using the DRM interface for it is another separate problem, and probably one that needs to be handled by the display server. Same problem exists for all DRM capable drivers.

File: hwdb.d/60-seat.hwdb (hunk 5 above)
Comment (hramrach) [reply hwdb.d/60-seat.hwdb:17]: Nevermind, HyperV was only mentioned in the bugreport as an example of another thing that is also broken.
